rbd-nbd: generate and send device cookie with netlink connect request

pkalever opened the pull request:
Based on the discussion at https://github.com/ceph/ceph/pull/40809, this PR adds below options for rbd-nbd cookie generation and management

* generate and send device cookie with netlink connect request

```
[root@linux-vm1]# rbd-nbd map rbd-pool/image0 --try-netlink
/dev/nbd0

[root@linux-vm1]# cat /sys/block/nbd0/backend
c704cb91-c6cf-466e-a335-0e935c0d5e47
```

* provide a flag to show device cookie with map command

```
[root@linux-vm1]# rbd-nbd map rbd-pool/image0 --try-netlink --show-cookie
/dev/nbd0 c704cb91-c6cf-466e-a335-0e935c0d5e47
```

* mandate device cookie for attach command

```
[root@linux-vm1]# rbd-nbd attach rbd-pool/image0 --device /dev/nbd0 --cookie c704cb91-c6cf-466e-a335-0e935c0d5e47
/dev/nbd0
```

* provide a way to show per device cookie with list command

```
[root@linux-vm1]# rbd-nbd list-mapped --all
id    pool      namespace  image   snap  device     cookie
8133  rbd-pool             image0  -     /dev/nbd0  c704cb91-c6cf-466e-a335-0e935c0d5e47
```

Needs: https://lkml.org/lkml/2021/4/29/274
Credits: To @idryomov and @trociny for all the discussions around this topic.
Signed-off-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>

trociny reviewed (commented):
I think it is important to provide some reasonable behavior for the case when the kernel does not support cookie, because I expect we will have a long period when this will be a common situation.

I suppose we can't just forbid `attach` for this case, because users may have already been using it. Also there may be a third party software that has already been using it and will break after rbd-nbd upgrade.

Making `--cookie` optional would resolve this or may be it will just delay the issue if we want to make it mandatory eventually. I have no a good idea about the solution. Any thoughts?

And I am not sure I like `--all` option for the list command. I was thinking we would just make list always print it. Actually I don't have a strong opinion. I see some advantages in printing it only with `--all`, e.g. we will still have the same list format by default for all devices. But if we want the `--all` option, we will also need to add it `rbd device list`.

Also I think you will need to update `qa/workunits/rbd/rbd-nbd.sh` and make sure it will work for both new and old kernels.

pkalever commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1326,6 +1352,8 @@ static int netlink_connect(Config *cfg, struct nl_sock *sock, int nl_id, int fd,
   NLA_PUT_U64(msg, NBD_ATTR_SERVER_FLAGS, flags);
   NLA_PUT_U64(msg, NBD_ATTR_DEAD_CONN_TIMEOUT, cfg->reattach_timeout);
+  if (!cfg->cookie.empty())
+    NLA_PUT_STRING(msg, NBD_ATTR_BACKEND_IDENTIFIER, cfg->cookie.c_str());
```
The attribute will be ignored.

pkalever commented:
> I think it is important to provide some reasonable behavior for the case when the kernel does not support cookie, because I expect we will have a long period when this will be a common situation.

With Kernel Changes:
-------------------
```
[root@linux-vm1 build]# uname -a
Linux linux-vm1 5.12.0+ #7 SMP Wed Apr 28 12:36:15 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

[root@linux-vm1 build]# ./bin/rbd-nbd map rbd-pool/image0 --try-netlink --show-cookie
/dev/nbd0 38f1d4ae-797e-49ce-acc5-e296279bb190

[root@linux-vm1 build]# cat /sys/block/nbd0/backend
38f1d4ae-797e-49ce-acc5-e296279bb190

[root@linux-vm1 build]# ./bin/rbd-nbd list-mapped --all
id    pool      namespace  image   snap  device     cookie                              
8232  rbd-pool             image0  -     /dev/nbd0  38f1d4ae-797e-49ce-acc5-e296279bb190

[root@linux-vm1 build]# pkill -9 rbd-nbd

[root@linux-vm1 build]# ./bin/rbd-nbd attach rbd-pool/image0 --device /dev/nbd0 --cookie 38f1d4ae-797e-49ce-acc5-e296279bb190
/dev/nbd0

[root@linux-vm1 build]# ./bin/rbd-nbd list-mapped --all
id    pool      namespace  image   snap  device     cookie                              
8347  rbd-pool             image0  -     /dev/nbd0  38f1d4ae-797e-49ce-acc5-e296279bb190

[root@linux-vm1 build]# ps -aux | grep rbd
root        8347  0.7  0.8 1565804 98300 pts/1   Sl   06:01   0:00 ./bin/rbd-nbd attach rbd-pool/image0 --device /dev/nbd0 --cookie 38f1d4ae-797e-49ce-acc5-e296279bb190
root        8374  0.0  0.0 221432   772 pts/1    S+   06:01   0:00 grep --color=auto rbd

[root@linux-vm1 build]# cat /sys/block/nbd0/backend
38f1d4ae-797e-49ce-acc5-e296279bb190
```

Without Kernel Changes:
----------------------
```
[root@linux-vm1 build]# uname -a
Linux linux-vm1 5.11.15-200.fc33.x86_64 #1 SMP Fri Apr 16 13:41:20 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

[root@linux-vm1 build]# ./bin/rbd-nbd map rbd-pool/image0 --try-netlink --show-cookie   <--  show-cookie is provided, if its not given then it should have behaved like earlier
/dev/nbd0 c3d2db83-8269-4f0e-b3c2-14aa2c4cfac5

[root@linux-vm1 build]# cat /sys/block/nbd0/backend
cat: /sys/block/nbd0/backend: No such file or directory

[root@linux-vm1 build]# ./bin/rbd-nbd list-mapped --all  <-- list will read it again from /sys/block/nbdX/backend and find it absent
id    pool      namespace  image   snap  device     cookie
8028  rbd-pool             image0  -     /dev/nbd0  -     

[root@linux-vm1 build]# pkill -9 rbd-nbd

[root@linux-vm1 build]# ./bin/rbd-nbd attach rbd-pool/image0 --device /dev/nbd0                                         
rbd-nbd: must specify cookie to attach

[root@linux-vm1 build]# ./bin/rbd-nbd attach rbd-pool/image0 --device /dev/nbd0 --cookie 1e524d06-5e99-48b3-aa4c-00c822aec4d2  <-- Please note the cookie provided here is a fake one as the code doesn't care about this, may be we can give a provision to allow attempting to attach when kernel cookie is absent?
/dev/nbd0

[root@linux-vm1 build]# ./bin/rbd-nbd list-mapped --all
id    pool      namespace  image   snap  device     cookie
8102  rbd-pool             image0  -     /dev/nbd0  -     

[root@linux-vm1 build]# ps -aux | grep rbd
root        8102  0.2  0.8 1565808 98308 pts/1   Sl   06:08   0:00 ./bin/rbd-nbd attach rbd-pool/image0 --device /dev/nbd0 --cookie 38f1d4ae-797e-49ce-acc5-e296279bb190
root        8133  0.0  0.0 221432   780 pts/1    S+   06:09   0:00 grep --color=auto rbd

[root@linux-vm1 build]# cat /sys/block/nbd0/backend
cat: /sys/block/nbd0/backend: No such file or directory
```

> 
> I suppose we can't just forbid `attach` for this case, because users may have already been using it. Also there may be a third party software that has already been using it and will break after rbd-nbd upgrade.
> 
> Making `--cookie` optional would resolve this or may be it will just delay the issue if we want to make it mandatory eventually. I have no a good idea about the solution. Any thoughts?

Maybe we can allow `attach` when the kernel doesn't have cookie support, like check for /sys/block/nbdX/backend if file doesn't exists allow `attach` without `--cookie` option?

> 
> And I am not sure I like `--all` option for the list command. I was thinking we would just make list always print it. Actually I don't have a strong opinion. I see some advantages in printing it only with `--all`, e.g. we will still have the same list format by default for all device types. But if we want the `--all` option, we will also need to add it `rbd device list`.

I don't have a strong opinion either, was just implementing it as per our discussion from the previous PR :-)

> 
> Also I think you will need to update `qa/workunits/rbd/rbd-nbd.sh` and make sure it will work for both new and old kernels.

I will check on what needs to be updated in `qa/workunits/rbd/rbd-nbd.sh`, and yes this patch works for older kernels too as mentioned before.

Thanks for your thoughts @trociny

pkalever commented:
And with the latest V2 patches, the changes respect backward compatibility and works with old and new kernels:

```
With Kernel Changes:
-------------------

[root@linux-vm1 build]# uname -a
Linux linux-vm1 5.12.0+ #7 SMP Wed Apr 28 12:36:15 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

[root@linux-vm1 build]# ./bin/rbd-nbd map rbd-pool/image0 --try-netlink --show-cookie
/dev/nbd0 d76226d0-be26-4b4d-a531-5ad58796a6c8

[root@linux-vm1 build]# ./bin/rbd-nbd list-mapped --all
id    pool      namespace  image   snap  device     cookie                              
7991  rbd-pool             image0  -     /dev/nbd0  d76226d0-be26-4b4d-a531-5ad58796a6c8

[root@linux-vm1 build]# pkill -9 rbd-nbd

[root@linux-vm1 build]# ./bin/rbd-nbd attach rbd-pool/image0 --device /dev/nbd0
rbd-nbd: must specify cookie to attach

[root@linux-vm1 build]# ./bin/rbd-nbd attach rbd-pool/image0 --device /dev/nbd0 --cookie d76226d0-be26-4b4d-a531-5ad58796a6c8
/dev/nbd0

[root@linux-vm1 build]# ps -aux | grep rbd
root        8066  0.4  0.7 1565800 97868 pts/1   Sl   08:21   0:00 ./bin/rbd-nbd attach rbd-pool/image0 --device /dev/nbd0 --cookie d76226d0-be26-4b4d-a531-5ad58796a6c8
root        8092  0.0  0.0 221432   836 pts/1    S+   08:21   0:00 grep --color=auto rbd




Without Kernel Changes:
----------------------

[root@linux-vm1 build]# uname -a
Linux linux-vm1 5.11.15-200.fc33.x86_64 #1 SMP Fri Apr 16 13:41:20 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

[root@linux-vm1 build]# ./bin/rbd-nbd map rbd-pool/image0 --try-netlink
/dev/nbd0

[root@linux-vm1 build]# ./bin/rbd-nbd list-mapped --all
id    pool      namespace  image   snap  device     cookie
8006  rbd-pool             image0  -     /dev/nbd0  -     

[root@linux-vm1 build]# pkill -9 rbd-nbd

[root@linux-vm1 build]# ./bin/rbd-nbd attach rbd-pool/image0 --device /dev/nbd0
/dev/nbd0

[root@linux-vm1 build]# ./bin/rbd-nbd list-mapped --all
id    pool      namespace  image   snap  device     cookie
8080  rbd-pool             image0  -     /dev/nbd0  -     

[root@linux-vm1 build]# ps -aux | grep rbd
root        8080  0.5  0.8 1565696 98656 pts/1   Sl   07:38   0:00 ./bin/rbd-nbd attach rbd-pool/image0 --device /dev/nbd0
root        8107  0.0  0.0 221432   780 pts/1    S+   07:38   0:00 grep --color=auto rbd
```

lxbsz commented:
...
> 
> > I suppose we can't just forbid `attach` for this case, because users may have already been using it. Also there may be a third party software that has already been using it and will break after rbd-nbd upgrade.
> > Making `--cookie` optional would resolve this or may be it will just delay the issue if we want to make it mandatory eventually. I have no a good idea about the solution. Any thoughts?
> 
> Maybe we can allow `attach` when the kernel doesn't have cookie support, like check for /sys/block/nbdX/backend if file doesn't exists allow `attach` without `--cookie` option?
> 

Maybe we should forbid `attach` as default ? And add one `--force` option and print some warning at the same time, which means risky, to let user to choose what to do ?

pkalever commented:
> > Maybe we can allow `attach` when the kernel doesn't have cookie support, like check for /sys/block/nbdX/backend if file doesn't exists allow `attach` without `--cookie` option?
> 
> Maybe we should forbid `attach` as default ? And add one `--force` option and print some warning at the same time, which means risky, to let user to choose what to do ?

That will break the existing users Xiubo, IMHO which is something that we do not want.

The latest changes here behave as usual when kernel changes are not available, and when kernel changes are available, they demand a cookie at `attach`.

Thanks!

lxbsz commented:
Hi Prasanna,

As we discussed, maybe we should do:

1, use the `<poolX>/<imageX>` to generate one uuid or one md5sum number instead of a random uuid.
2, so the `rbd-nbd` tool could check and compare the new `<poolY>/<imageY>` with the saved number in `/sys/block/nbdX/backend` when `attaching` to verify whether they match. No need to store the random uuid by the userspace apps for each image.

pkalever commented:
> Hi Prasanna,
> 
> As we discussed, maybe we should do:
> 
> 1, use the `<poolX>/<imageX>` to generate one uuid or one md5sum number instead of a random uuid.
> 2, so the `rbd-nbd` tool could check and compare the new `<poolY>/<imageY>` with the saved number in `/sys/block/nbdX/backend` when `attaching` to verify whether they match. No need to store the random uuid by the userspace apps for each image.

Hello Xiubo,
I like the idea because it makes users free from maintaining the cookies per image.
But I'm just thinking about how this works in cases like live migration(moving images between pools)?

Example:
$ rbd-nbd map pool1/image1 --try-netlink --reattach-timeout 600
/dev/nbd0
$ pkill -9 rbd-nbd

Move the image to a different pool
$ rbd migration prepare pool1/image1 pool2/image1
$ rbd migration execute pool2/image1
$ rbd migration commit pool2/image1

Now claim the same device, but for the migrated destination (target) image, on the same host node
$ rbd-nbd attach pool2/image1 --device /dev/nbd0

In this case, one needs to supply the source image_spec's(pool1/image1) md5sum cookie manually.

Now, the question is, do we want to give two ways?
1. rbd-nbd generate the cookie (md5sum) of image_spec internally and try to attach without the user providing `--cookie`
2. If the internally generated cookie doesn't match at kernel (/sys/block/nbdX/backend), the attach will bailout and the user should give the `--cookie` manually.

Thinking about automation scripts, they still have to save the cookie and keep it ready at least to address this special case.

Maybe @idryomov @trociny might add more thoughts here?

Thanks!

trociny commented:
> > Hi Prasanna,
> > As we discussed, maybe we should do:
> > 1, use the `<poolX>/<imageX>` to generate one uuid or one md5sum number instead of a random uuid.
> > 2, so the `rbd-nbd` tool could check and compare the new `<poolY>/<imageY>` with the saved number in `/sys/block/nbdX/backend` when `attaching` to verify whether they match. No need to store the random uuid by the userspace apps for each image.
> 
> Hello Xiubo,
> I like the idea because it makes users free from maintaining the cookies per image.
> But I'm just thinking about how this works in cases like live migration(moving images between pools)?

Yes it will not work for migration. That is why we need something different than image spec for the identifier. 

> Example:
> $ rbd-nbd map pool1/image1 --try-netlink --reattach-timeout 600
> /dev/nbd0
> $ pkill -9 rbd-nbd

Don't use `-9` for detaching. Give the rbd-nbd a chance to flush pending io before termination. Use the default TERM signal or bette `rbd-nbd detach`.

> Move the image to a different pool
> $ rbd migration prepare pool1/image1 pool2/image1

Actually you may/should attach after this step. This is the main reason why we have migration in several steps -- to make the downtime small.

> $ rbd migration execute pool2/image1
> $ rbd migration commit pool2/image1
> 
> Now claim the same device, but for the migrated destination (target) image, on the same host node
> $ rbd-nbd attach pool2/image1 --device /dev/nbd0
> 
> In this case, one needs to supply the source image_spec's(pool1/image1) md5sum cookie manually.
> 
> Now, the question is, do we want to give two ways?
> 
> 1. rbd-nbd generate the cookie (md5sum) of image_spec internally and try to attach without the user providing `--cookie`
> 2. If the internally generated cookie doesn't match at kernel (/sys/block/nbdX/backend), the attach will bailout and the user should give the `--cookie` manually.

I don't think it is necessary, as it does not improve it radically, but I would not object.

lxbsz commented:
> > > Hi Prasanna,
> > > As we discussed, maybe we should do:
> > > 1, use the `<poolX>/<imageX>` to generate one uuid or one md5sum number instead of a random uuid.
> > > 2, so the `rbd-nbd` tool could check and compare the new `<poolY>/<imageY>` with the saved number in `/sys/block/nbdX/backend` when `attaching` to verify whether they match. No need to store the random uuid by the userspace apps for each image.
> > 
> > 
> > Hello Xiubo,
> > I like the idea because it makes users free from maintaining the cookies per image.
> > But I'm just thinking about how this works in cases like live migration(moving images between pools)?
> 
> Yes it will not work for migration. That is why we need something different than image spec for the identifier.
> 

Okay, I didn't notice the migration case.

trociny commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1718,6 +1747,13 @@ static int do_map(int argc, const char *argv[], Config *cfg, bool reconnect)
+    if (!reconnect) {
+      /* FIXME: If kernel version doesn't suppot NBD_ATTR_BACKEND_IDENTIFIER
+       * we should skip generating uuid for cookie
+       */
```
nit: This comment does not look useful to me. As I understand we are going just to always try to set the cookie. Do you really plan to add a check if the kernel support NBD_ATTR_BACKEND_IDENTIFIER?

pkalever replied:
At some point I thought there is a need for it, but now I don't think we need one.
I will remove the commnet.

pkalever replied:
At some point, I thought there is a need for it, but now I don't think we need one.
I will remove the commnet.

trociny commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -2104,6 +2164,16 @@ static int parse_args(vector<const char*>& args, std::ostream *err_msg,
+        if (access(path.c_str(), F_OK) == 0) {
+          *err_msg << "rbd-nbd: must specify cookie to attach";
+          return -EINVAL;
+        }
```
nit: I would make `get_cookie` function, which could be used both here and when listing.

pkalever replied:
Sure.

trociny commented on qa/workunits/rbd/rbd-nbd.sh:
```diff
@@ -331,4 +331,23 @@ _sudo dd if=${DATA} of=${TEMPDIR}/mnt/test1 bs=1M count=1 oflag=direct
+_sudo mount ${DEV[0]} ${TEMPDIR}/mnt
+_sudo rbd-nbd detach ${POOL}/${IMAGE}
+expect_false get_pid
+_sudo rbd-nbd attach --device ${DEV[0]} ${POOL}/${IMAGE} --cookie ${DEV[1]}
```
I suppose when running this on an old kernel, `${DEV[1]}` will be empty and the test will fail?

Could you improve this. Alternatively, may be we should allow empty '--cookie' param, for cases like this?

pkalever replied:
oops! I would prefer to add a condition for empty string check for ${DEV[1]} and at accordingly.
Because there is `--force` option for old kernel versions now (This patch needs a refresh to use latest --force option)
But you can correct me, I don't have a strong opinion here.

trociny commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -2049,10 +2100,19 @@ static int parse_args(vector<const char*>& args, std::ostream *err_msg,
+    } else if (ceph_argparse_flag(args, i, "--show-cookie", (char *)NULL)) {
+      cfg->show_cookie = true;
+    } else if (ceph_argparse_witharg(args, i, &cfg->cookie, "--cookie", (char *)NULL)) {
+      if (!is_valid_uuid(cfg->cookie)) {
```
nit: Actually I don't see much value in this check. If it is not valid it will fail later anyway. But I don't mind much, you can leave it as it is now if other people think it is ok.

pkalever replied:
Sure, will take care in next respin.

idryomov replied:
Yeah, I'd drop this check and `is_valid_uuid()`.

pkalever replied:
Sure

trociny reviewed (commented):
Just a note for discussion. We could allow to specify `--cookie` on `map` command, so a user could set it to any value she likes. Not sure we need this functionality, it might be useful for some cases, but seems to introduce yet another chance to shoot in the foot.

pkalever commented:
> Just a note for discussion. We could allow to specify `--cookie` on `map` command, so a user could set it to any value she likes. Not sure we need this functionality, it might be useful for some cases, but seems to introduce yet another chance to shoot in the foot.

I agree it is useful for some cases, I'm happy to provide the changes. But I will wait to hear @idryomov thoughts on this before making the changes.

Thanks @trociny for the review.

pkalever commented:
Hello @idryomov @trociny

The kernel patch adding a way to identify device backends is applied and should be part of kernel version 5.14
I have respin this series by addressing previous review comments from @trociny also rebased this on the current master branch. Please take a look.

I'm yet to work on enabling users to specify --cookie on map command, this is something I'm waiting on @idryomov to decide.

Many Thanks!

pkalever commented:
@idryomov @trociny Just in case if this got lost, here is an attempt to remind you about the review, please help.

Thanks!

trociny commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -2169,21 +2169,24 @@ static int parse_args(vector<const char*>& args, std::ostream *err_msg,
+      // NBD_ATTR_BACKEND_IDENTIFIER support for compatibility
+      cookie = get_cookie(cfg->devpath);
+      if (!cookie.empty()) {
+	if (cfg->cookie.empty()) {
```
nit: is it just github that shows indentation wrong or is it wrong indeed?

trociny commented on src/test/cli/rbd/help.t:
```diff
@@ -625,13 +625,14 @@
-  usage: rbd device list [--device-type <device-type>] [--format <format>] 
-                         [--pretty-format] 
+  usage: rbd device list [--device-type <device-type>] [--options <options>] 
+                         [--format <format>] [--pretty-format] 
```
I think I would prefer if we just add `--all` option instead of introducing device specific `--options`. For the `list` command it looks awkward.

pkalever replied:
Sure, I'm fine with it.

trociny reviewed (commented):
Otherwise LGTM

pkalever reviewed (commented):
Thanks @trociny, Had addressed the review comments, PTAL.

trociny reviewed (approved):
LGTM.

BTW, what it is the status of the kernel patch? Just wondering if there a way to try it already without building a custom kernel...

pkalever commented:
> LGTM.
> 
> BTW, what it is the status of the kernel patch? Just wondering if there a way to try it already without building a custom kernel...

Thanks for the review @trociny 

AFAIK It should land in the 5.14 version. And I dought we will have anything ready to consume this early.
I think @idryomov is an expert there, Ilya do you know if we have anything to consume including https://lkml.org/lkml/2021/6/16/739 ?

pkalever commented:
Hey @idryomov 

Please help take this forward.

Thanks!

pkalever commented:
@idryomov just to remind, it appears that this is waiting on you, could you please help review it.

idryomov commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -928,6 +940,27 @@ class NBDListIterator {
+{
+  std::string cookie;
+
+  if (devpath.empty()) {
```
Can `devpath` actually be empty here?

idryomov commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1761,7 +1812,10 @@ static int do_map(int argc, const char *argv[], Config *cfg, bool reconnect)
 
-    cout << cfg->devpath << std::endl;
+    if (use_netlink && cfg->show_cookie && !cfg->cookie.empty())
+      cout << cfg->devpath << " " << get_cookie(cfg->devpath) << std::endl;
```
Might be good to `ceph_assert` that `get_cookie()` returns exactly `cfg->cookie`.

pkalever replied:
Sure.

idryomov replied:
Isn't that done only on attach?  The assert I was envisioning would ensure that the cookie that we generate on map and pass to the kernel is returned through `get_cookie()` intact.

pkalever replied:
Right, but that was intentional.
Having an assertion here will make us future-proof :+1:

idryomov commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1912,14 +1969,23 @@ static int do_list_mapped_devices(const std::string &format, bool pretty_format)
-          << cfg.snapname << cfg.devpath << TextTable::endrow;
+      if (cfg.cookie.empty()) {
+        cfg.cookie = "-";
+      }
```
Let's report an empty cookie as empty (i.e. don't munge it to `-`).  If we add support for custom user-provided cookies in the future, special casing `-` would be awkward.

pkalever replied:
sure, I just followed the `snapname` just above.

idryomov commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -2101,12 +2176,25 @@ static int parse_args(vector<const char*>& args, std::ostream *err_msg,
+      // Allowing attach without --cookie option for kernel without
+      // NBD_ATTR_BACKEND_IDENTIFIER support for compatibility
+      cookie = get_cookie(cfg->devpath);
+      if (!cookie.empty()) {
```
I think this is wrong.  While `rbd-nbd attach` may allow attaching without `--cookie`, `rbd device attach` should not.
`rbd device attach` semantics should be:
- if no `--cookie`, require `--force` (already there)
- if `--cookie`, fail for kernels without the patch (seems to be missing)

pkalever replied:
I don't think map displays the cookie even if `--show-cookie` is mentioned when the kernel changes are missing. So wondering from where do they even supply the cookie from at the time of attach?

If it is just to defend in the code, I can add one.

idryomov replied:
It is to protect the (sometimes clueless) user from shooting themselves in the foot.  Consider:
- argh, the command without `--cookie` doesn't work, wants `--force`
- but `--force` sounds dangerous, especially in a script (non-interactive)...
- what if I pass `--cookie 123` instead?
- it works!

Also, in the future we might want to support user-supplied cookies, passed on map.  In that case the cookie could be stored somewhere (or they could choose to use the image name, etc).

pkalever replied:
OK I hear that as an ack for user-supplied cookie :-)
I can add a quick change.

idryomov commented on qa/workunits/rbd/rbd-nbd.sh:
```diff
@@ -365,22 +365,31 @@ cat ${LOG_FILE}
 # test detach/attach
-DEV=`_sudo rbd device --device-type nbd --options try-netlink map ${POOL}/${IMAGE}`
+OUT=`_sudo rbd device --device-type nbd --options try-netlink,show-cookie map ${POOL}/${IMAGE}`
+read -a DEV <<< "${OUT}"
```
Nit: why array and not something like `read DEV COOKIE <<< "$OUT"`?  `--cookie $COOKIE` would read much nicer than `--cookie ${DEV[0]}`.

pkalever replied:
No preference, will update.

idryomov commented:
> AFAIK It should land in the 5.14 version. And I dought we will have anything ready to consume this early.
> I think @idryomov is an expert there, Ilya do you know if we have anything to consume including https://lkml.org/lkml/2021/6/16/739 ?

We can use our "testing" kernel -- the one we are testing the kernel client with.  It is currently based on 5.14-rc2, so the patch is there.  (Using the "testing" kernel might also help with https://tracker.ceph.com/issues/50905 and alike which seem to pop up on 5.4.)

pkalever commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -163,8 +166,11 @@ static void usage()
             << "  --reattach-timeout <sec>      Set nbd re-attach timeout\n"
             << "                                (default: " << Config().reattach_timeout << ")\n"
             << "  --try-netlink                 Use the nbd netlink interface\n"
+            << "  --show-cookie                 Show device cookie\n"
```
Makesense, done

pkalever commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -928,6 +940,27 @@ class NBDListIterator {
   }
 };
 
+static std::string get_cookie(const std::string devpath)
```
Argh, will fix it

pkalever commented on src/test/cli/rbd/help.t:
```diff
@@ -624,13 +625,14 @@
   
   Optional arguments
     -t [ --device-type ] arg device type [ggate, krbd (default), nbd]
+    --all                    show more device specific details
```
I still feel breaking the compatibility is not good. Maybe in the future once the kernel version is widely used we can switch to `--all` as default (meaning show cookie by default). `--all`  can be kept reserved for any new additional info like cookie now? But that's just me.

idryomov replied:
The way I see it, there is no compatibility to break here.  We have provided structured output for many years, long before `rbd-nbd` appeared.  If someone still insists on scraping the plain text table, they should update their scripts/tools.

pkalever replied:
OK. I shall drop the changes both in rbd and rbd-nbd CLI covering `--all` option. Rather I will make the behaviour default i.e show cookie with `list` directly.

pkalever reviewed (commented):
Thanks @idryomov, updating the PR with accepted review comments.

pkalever commented:
@idryomov @trociny could you please help review this? thanks!

idryomov commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1761,7 +1797,16 @@ static int do_map(int argc, const char *argv[], Config *cfg, bool reconnect)
+      cookie = get_cookie(cfg->devpath);
+    }
+    if (!cookie.empty()) {
+      ceph_assert(cookie.compare(cfg->cookie) == 0);
```
Nit: the assert condition could be just `cookie == cfg->cookie`.

idryomov commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1557,6 +1588,7 @@ static int do_map(int argc, const char *argv[], Config *cfg, bool reconnect)
   unsigned long size;
   unsigned long blksize = RBD_NBD_BLKSIZE;
   bool use_netlink;
+  uuid_d uuid_gen;
```
Nit: move `uuid_gen` into the (only) `if` block where it is used.

idryomov commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1761,7 +1797,16 @@ static int do_map(int argc, const char *argv[], Config *cfg, bool reconnect)
-    cout << cfg->devpath << std::endl;
+    std::string cookie;
+    if (use_netlink && cfg->show_cookie && !cfg->cookie.empty()) {
+      cookie = get_cookie(cfg->devpath);
```
It would be nice to assert on the returned cookie independent of `--show-cookie`.  Could you restructure these two `if` statements so that `get_cookie()` is called and verified on `use_netlink && !cfg->cookie.empty()`?

pkalever commented:
@idryomov addressed all the review comments.

Tested this series with the below cases:
```
map -> cookie     | show-cookie
map -> no-cookie  | show-cookie
map -> cookie     | no-show-cookie
map -> no-cookie  | no-show-cookie

attach -> cookie     | show-cookie
attach -> no-cookie  | show-cookie
attach -> cookie     | no-show-cookie
attach -> no-cookie  | no-show-cookie
```

Validated all the above cases with and without kernel change using rbd and rbd-nbd CLI's.

trociny commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1793,9 +1794,12 @@ static int do_map(int argc, const char *argv[], Config *cfg, bool reconnect)
+      // cookie might be empty if kernel doesn't have support
+      if (!cookie.empty()) {
+        ceph_assert(cookie == cfg->cookie);
+      }
```
nit: `ceph_assert(cookie == cfg->cookie || cookie.empty());`

trociny reviewed (commented):
Otherwise LGTM

pkalever commented on qa/workunits/rbd/rbd-nbd.sh:
```diff
@@ -365,22 +365,39 @@ cat ${LOG_FILE}
+    get_pid ${POOL}
+    test "${ANOTHER_COOKIE}" = "${COOKIE}"
+fi
+unmap_device ${DEV} ${PID}
```
Done now @trociny.
Thanks!

pkalever commented:
@idryomov could you please help here with your review and approval.

Thanks!

pkalever commented:
@idryomov ping

pkalever commented:
@idryomov can you please review this?

idryomov commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -789,6 +795,7 @@ class NBDListIterator {
 
       *cfg = Config();
       cfg->devpath = "/dev/nbd" + stringify(m_index++);
+      cfg->cookie = get_cookie(cfg->devpath);
```
Why are we querying the kernel for the cookie here?  It looks like the config (along with the cookie) may already be cached in `m_mapped_info_cache`.  Also, if we do it in `get_mapped_info()`, `find_attached()` wouldn't need a cookie parameter.  I would move this call to line 912 and possibly validate that the cookie from `parse_args()` (if any) matches the one from the kernel.

pkalever replied:
looks reasonable to move it under get_mapped_info()

idryomov commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -2102,12 +2148,28 @@ static int parse_args(vector<const char*>& args, std::ostream *err_msg,
+        if (cfg->cookie.empty()) {
+          *err_msg << "rbd-nbd: must specify cookie to attach";
+          return -EINVAL;
+        } else if (cookie.compare(cfg->cookie) != 0) {
```
Nit: cookie != cfg->cookie

idryomov commented on qa/workunits/rbd/rbd-nbd.sh:
```diff
@@ -365,22 +365,39 @@ cat ${LOG_FILE}
 unmap_device ${DEV} ${PID}
+# if kernel supports cookies
+if [ -n "${COOKIE}" ]; then
+    OUT=`_sudo rbd device --device-type nbd --show-cookie --cookie="${COOKIE}" --options try-netlink map ${POOL}/${IMAGE}`
```
Nit: I would pass a literal string such as abcde for cookie here to show that it doesn't have to be a UUID.

pkalever commented:
@idryomov Thanks for the review, I have updated the series addressing your review comments and rebased to later master.

idryomov commented on qa/workunits/rbd/rbd-nbd.sh:
```diff
@@ -365,22 +365,39 @@ cat ${LOG_FILE}
 unmap_device ${DEV} ${PID}
+# if kernel supports cookies
+if [ -n "${COOKIE}" ]; then
+    OUT=`_sudo rbd device --device-type nbd --show-cookie --cookie="abcde" --options try-netlink map ${POOL}/${IMAGE}`
```
I was going to say "make it look like other options (i.e. get rid of `=` and quotes)" here but the quotes made me realize that it would be good to test a cookie with a space in it.  Let's add a space:
```suggestion
    OUT=`_sudo rbd device --device-type nbd --show-cookie --cookie "abc de" --options try-netlink map ${POOL}/${IMAGE}`
```

pkalever replied:
@idryomov why would someone include spaces in cookie?
I expect the code to assert at `ceph_assert(cookie == cfg->cookie || cookie.empty());` as get_cookie() will return only first part of the cookie i.e `abc` and not `abc de`.

Should I adjust get_cookie() such a way that a complete string (including spaces) is returned? I don't know why would someone sanely include a space in cookie though.

idryomov commented on qa/workunits/rbd/rbd-nbd.sh:
```diff
@@ -365,22 +365,39 @@ cat ${LOG_FILE}
+    OUT=`_sudo rbd device --device-type nbd --show-cookie --cookie="abcde" --options try-netlink map ${POOL}/${IMAGE}`
+    read DEV ANOTHER_COOKIE <<< "${OUT}"
+    get_pid ${POOL}
+    test "${ANOTHER_COOKIE}" = "abcde"
```
```suggestion
    test "${ANOTHER_COOKIE}" = "abc de"
```
